deltacat/utils/metrics.py

```diff
@@ -0,0 +1,108 @@
+import ray
+import urllib
+import logging
+import yaml
+
+from deltacat import logs
+from enum import Enum
+from typing import Dict, Any, List, Callable
+from deltacat.aws.clients import resource_cache
+
+from botocore.exceptions import ClientError
+
+logger = logs.configure_deltacat_logger(logging.getLogger(__name__))
+
+DEFAULT_DELTACAT_METRICS_NAMESPACE = "ray-deltacat-profiling"
+
+
+class MetricsType(str, Enum):
+    TIMER = "timer"
+
+
+class MetricsDimensionType(str, Enum):
+    NODE_IP = "node_ip"
+    RAY_TASK_ID = "task_id"
+    RAY_WORKER_ID = "worker_id"
+
+
+def _build_metrics_name(metrics_type: Enum, metrics_name: str) -> str:
+    metrics_name_with_type = f"{metrics_name}_{metrics_type}"
+    return metrics_name_with_type
+
+
+def _get_current_node_ip() -> str:
+    current_node_ip = urllib.request.urlopen(
+        "http://169.254.169.254/latest/meta-data/local-ipv4"
+    ).read().decode("utf-8")
+    return {"Name": f"node_ip",
+            "Value": f"{current_node_ip}"
+            }
```

Comment:
Same here above dependency on platform. Can we use a cloudwatch metrics config or environment variables to resolve these? For example, take method `_get_current_node_ip` as argument as `metrics_config` of type 'CloudWatch'?

Reply:
Not sure I understand the approach here, could you please elaborate more on the part `as metrics_config of type 'CloudWatch'`?

Reply:
The idea is you could take an optional argument to `compact_partition` method named `metrics_config` with below signature:

```python
@dataclass
class MetricsConfig:
  def __init__(self, type: str, metrics_dimensions_provider: Callable, cluster_name_provider: Callable):
    self.type = type
    self.retrieve_metrics_dimensions = retrieve_metrics_dimensions
    self.cluster_name_provider = cluster_name_provider

# then in the driver we could,

def metrics_dimensions_callable():
  node_ip = _get_current_node_id()
  instance_region = _get_instance_region()
  # etc etc.

def cluster_name_provider():
  # provide cluster name

config = MetricsConfig(MetricsConfigType.CloudWatch, metrics_dimensions_callable, cluster_name_provider)

compact_partition(.., metrics_config=config, ...)

```

Reply:
Do we really need node ip? for debugging purpose, probably node id is sufficient? What we want is probably just an identifier to locate the actual worker logs, e.g., worker id; Please share ur thoughts.

Reply:
Here the node id is not included. My original thought is that Ray uses node ip instead of node id  in their dashboard and logs, so using same node ip as identifier could make debugging easier for us?

Reply:
> 
Given it a second thought, if we wrap the MetricsConfig in driver, the node id can not be derived correctly because the driver do not know each individual node's ip beforehand?

Reply:
I am not sure I understood. Let's sync up offline.